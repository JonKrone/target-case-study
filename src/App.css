:root {
  --border-color: #4c5c68;
  --background-color: #dcdcdd;
}

nav li:nth-child(odd) {
  background-color: #c5c3c6;
}

button {
  border: lightgray;
  background: lightgray;
}

button:hover {
  cursor: pointer;
  background: #1985a1;
  border: #1985a1;
}
