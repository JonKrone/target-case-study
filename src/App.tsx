import React, { Suspense } from 'react'
import { Provider } from 'use-http'
import { BrowserRouter as Router, Route, Link } from 'react-router-dom'

import './App.css'
import StopList from './components/StopList'
import RouteList from './components/RouteList'

function App() {
  return (
    // It looks like use-https is not properly persisting multiple sequential calls
    // I'll be posting an issue in the project once I can replicate it in a codesandbox
    <Provider
      url="https://svc.metrotransit.org/NexTrip"
      options={{ mode: 'cors', persist: true }}
    >
      <Router>
        <div
          style={{
            display: 'flex',
            justifyContent: 'center',
            alignItems: 'center',
            marginTop: '10%',
          }}
        >
          <div
            style={{
              display: 'flex',
              flexDirection: 'column',
              width: 1000,
              height: 600,
              backgroundColor: 'white',
              border: '1px solid var(--border-color)',
              borderRadius: '5px',
            }}
          >
            <header style={{ height: '15%', textAlign: 'center' }}>
              <Link to="/" style={{ textDecoration: 'none' }}>
                <h1>NextTrip tips</h1>
              </Link>
            </header>

            <section
              style={{
                display: 'flex',
                height: '85%',
                borderTop: '2px solid #4c5c68',
              }}
            >
              <Suspense
                fallback={
                  <div style={{ textAlign: 'center', width: '100%' }}>
                    <h1>Loading Routes</h1>
                  </div>
                }
              >
                <RouteList />

                {/* New finding: you can describe multiple path matches */}
                <Route path={['/:route/:direction', '/']}>
                  <StopList />
                </Route>
              </Suspense>
            </section>
          </div>
        </div>
      </Router>
    </Provider>
  )
}

export default App

/**
 * Initial App structure thoughts
 *
 * container
 *  left sidebar Route and Route-Direction selector
 *    fetch available routes
 *    display all Routes with North/South buttons aside them
 *    when a route is selected, update the URL
 *  Right content area for Route-Stops
 *    when no route is selected
 *      display some cute waiting signal?
 *    when a route is selected
 *      fetch list of directions
 *      render that list
 *      Optional: add a link to a Google Map location for each stop
 *      Stretch: Render stops layed over a Google/Mapbox map widget
 *
 * To enable back/forward, we probably want to have Route and Route-Direction be a URL-parameter
 *
 * Stretch thought: We would probably get into NextTrip rate limiting if we were to pull all
 * Route-Stops for all Routes, so that we could overlay them all on top of a Map at once. But
 * only in a real high-volume (production) usage, so maybe we can get away with it for this
 * test app :)
 */
