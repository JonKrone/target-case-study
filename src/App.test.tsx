import React from 'react'
import { render } from '@testing-library/react'
import App from './App'

describe('App', () => {
  it('renders', () => {
    const { getByText, asFragment } = render(<App />)

    expect(getByText(/NextTrip tips/i)).toBeInTheDocument()
    expect(getByText(/Loading Routes/)).toBeInTheDocument()
    expect(getByText(/Please select/)).toBeInTheDocument()

    expect(asFragment()).toMatchSnapshot()
  })
})
