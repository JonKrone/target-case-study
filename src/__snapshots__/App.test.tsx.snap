// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders 1`] = `
<DocumentFragment>
  <div
    style="display: flex; justify-content: center; align-items: center; margin-top: 10%;"
  >
    <div
      style="display: flex; width: 1000px; height: 600px; flex-direction: column; background-color: white; border-radius: 5px;"
    >
      <header
        style="height: 15%; text-align: center;"
      >
        <a
          href="/"
          style="text-decoration: none;"
        >
          <h1>
            NextTrip tips
          </h1>
        </a>
      </header>
      <section
        style="display: flex; height: 85%; border-top: 2px solid #4c5c68;"
      >
        <nav
          style="width: 50%; overflow: auto; border-bottom-left-radius: 5px; display: none;"
        >
          <ul
            style="margin: 0px; padding-left: 0px; list-style: none;"
          />
        </nav>
        <section
          style="flex: 1; overflow-y: auto; border-bottom-right-radius: 5px; display: none;"
        >
          <h1
            style="text-align: center;"
          >
            Please select a route
          </h1>
        </section>
        <h1
          style="text-align: center; width: 100%;"
        >
          Loading Routes
        </h1>
      </section>
    </div>
  </div>
</DocumentFragment>
`;
